text/0000-dynamic-events.md

```diff
@@ -0,0 +1,148 @@
+
+### Directive behaviour 
+
+The `lwc:on` directive would accept an object. For each property of the object, it would add a event listener to the `element` that listens for the event type specified by property's key. The property's value with `this` set to the owner `component` would be used for handling the event. Consumer would not recieve any reference to the actual event handlers used and hence they can't remove the event listeners added by this directive.
+
+### Considered Properties
+
+Only own enumerable string-keyed properties of object passed to `lwc:on` will be considered. Other properties, i.e. inherited properties, non-enumerable properties and symbol-keyed properties shall be ignored. The only restriction on property keys is that they must be strings — any string, such as 'fooBar', 'foo-bar', or even '', is valid.
+
+### Caching
+
+#### For static components, i.e. components declared in template directly using its selector
+Since it is uncommon for event listeners to change after the start of the owner component's intial render, We can cache them when first encountered, to improve performance. Note that this is same as how `on{eventType}` in template HTML works currently. For consumers, the implication of this would be that any changes made to the object passed to `lwc:on` after the first `connectedCallback` of owner component would cause no effect.
+
+#### For dynamic components, i.e. components created using `lwc:component` directive
+Since it is uncommon for event listeners to change if the constructor passed to `lwc:is` doesn't change, We can skip patching of event listeners if the `element` doesn't change. For consumers, the implication of this would be that after the first `connectedCallback` of owner component, any changes made to the object passed to `lwc:on` would cause no effect until the constructor passed to `lwc:is` itself is changed - at which point all existing listeners are removed and new ones are added based on the current value of object passed to `lwc:on`.
+
+### Overriding
+
+There can be only one `lwc:on` directive on an element.  
+If `lwc:on` specifies a listener for the same event type as an event listener declared directly in the HTML template, only the listener from `lwc:on` will be applied, and the other listener will be ignored. In the example below, for `x-child`, only one listener for event type `foo` will be added and `childEventHandlers.foo` would be used for event handler.  
+
+```js
+// x/myComponent.js
+export default class MyComponent extends LightningElement {
+
+    childEventHandlers = {
+        foo: function (event) {console.log('lwc:on');} ,
+    };
+
+    fooHandler(){
+        console.log('onfoo');
+    }
+
+}
+```
+
+```html
+ <!-- x/myComponent.html -->
+<template>
+  <x-child lwc:on={childEventHandlers} onfoo={fooHandler}></x-child>

```

Comment:
As we talked about in our engineering discussion, the LWC team would like to avoid this scenario altogether. Here's what the behavior should be: if an `lwc:on` is provided on the same tag as `onfoo` or `onbar` or `onanything`, then an error is thrown at compilation. The message should be `It is not permitted to use declarative event listeners alongside lwc:on`.

In the future, we may revisit and relax the compile-time restriction. However, it seems like a component author using `lwc:on` should be able to accomplish everything they need to accomplish without also using `on*` declarative listeners. Without a compelling use case, we prefer the more restrictive approach.

Reply:
I'm not sure I buy the complexity argument in the other comment, but I'm supportive of starting conservatively and opening things up if there are valid use cases.